src/systems/physics/Physics.cc

```diff
@@ -1536,56 +1730,70 @@ void PhysicsPrivate::UpdateSim(EntityComponentManager &_ecm)
-            // The Pose component, _pose, of this link is the initial
-            // transform of the link w.r.t its model. This component never
-            // changes because it's "fixed" to the model. Instead, we change
-            // the model's pose here. The physics engine gives us the pose of
-            // this link relative to world so to set the model's pose, we have
-            // to post-multiply it by the inverse of the initial transform of
-            // the link w.r.t to its model.
-            auto mutableParentPose =
-              _ecm.Component<components::Pose>(_parent->Data());
-            *(mutableParentPose) = components::Pose(_pose->Data().Inverse() +
-                                           math::eigen3::convert(worldPose));
-            _ecm.SetChanged(_parent->Data(), components::Pose::typeId,
+            // This is the canonical link, update the top level model.
+            // The pose of this link w.r.t its top level model never changes
+            // because it's "fixed" to the model. Instead, we change
+            // the top level model's pose here. The physics engine gives us the
+            // pose of this link relative to world so to set the top level
+            // model's pose, we have to post-multiply it by the inverse of the
+            // transform of the link w.r.t to its top level model.
+            math::Pose3d linkPoseFromTopLevelModel;
+            linkPoseFromTopLevelModel =
+                this->RelativePose(topLevelModel, _entity, _ecm);
+
+            // update top level model's pose
+            auto mutableModelPose =
+               _ecm.Component<components::Pose>(topLevelModel);
+            *(mutableModelPose) = components::Pose(
+                math::eigen3::convert(worldPose) *
+                linkPoseFromTopLevelModel.Inverse());
+
+            _ecm.SetChanged(topLevelModel, components::Pose::typeId,
                 ComponentState::PeriodicChange);
           }
           else
           {
-            // Compute the relative pose of this link from the model
+            // Compute the relative pose of this link from the top level model
+            // first get the world pose of the top level model
+            auto worldComp =
+                _ecm.Component<components::ParentEntity>(topLevelModel);
+            // if the worldComp is a nullptr, something is wrong with ECS
+            if (!worldComp)
+            {
+              ignerr << "The parent component of " << topLevelModel
+                     << " could not be found. This should never happen!\n";
+              return true;
+            }
+            math::Pose3d parentWorldPose =
+                this->RelativePose(worldComp->Data(), _parent->Data(), _ecm);
+
+            // Note that if the canonical link is inside a nested model, the
+            // immediate parent model of this link will not have its pose
+            // updated. Nested models have a fixed pose relative to the top
```

Comment:
> Nested models have a fixed pose relative to the top level model.

I think this is only true for the nested model that has the canonical link, right? Maybe the comment can be updated to avoid confusion?

Reply:
actually this comment is kind of out-of-place as the code here is for mainly about non-canonical links. I've updated the comment in a71a3d2e2e65bab10243987f91ed238a1329f7ab